src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs

```diff
@@ -665,5 +662,69 @@ private static bool LogFailure(ValueStopwatch stopwatch)
+        {
+            // Queue the function to be invoked asynchronously.  Since this is doing synchronous work on a thread pool thread,
+            // we want to limit how many threads end up being blocked.  We could employ a semaphore to limit overall usage, but
+            // a common case is that DNS requests are made for only a handful of endpoints, and a reasonable compromise is to
+            // ensure that requests for a given host are serialized.  Once the data for that host is cached locally by the OS,
+            // the subsequent requests should all complete very quickly, and if the head-of-line request is taking a long time
+            // due to the connection to the server, we won't block lots of threads all getting data for that one host.  We also
+            // still want to issue the request to the OS, rather than having all concurrent requests for the same host share the
+            // exact same task, so that any shuffling of the results by the OS to enable round robin is still perceived.
+            return AsyncOverSyncThrottle<TResult>.RunAsync(func!, arg, cancellationToken);
+        }
+
+        /// <summary>Throttles async-over-sync tasks to queue one at a time per key.</summary>
+        /// <typeparam name="TResult">Specifies the result type of the function to be invoked asynchronously.</typeparam>
+        private static class AsyncOverSyncThrottle<TResult>
+        {
+            /// <summary>Mapping from key to current task in flight for that key.</summary>
+            private static readonly Dictionary<object, Task<TResult>> s_tasks = new Dictionary<object, Task<TResult>>();
+
+            /// <summary>Queues a task to invoke the specified function with the specified key.</summary>
+            public static async Task<TResult> RunAsync(Func<object, TResult> func, object key, CancellationToken cancellationToken)
+            {
+                // Create a task to invoke the function.
+                var task = new Task<TResult>(func!, key, cancellationToken, TaskCreationOptions.DenyChildAttach);
+                try
+                {
+                    // If there's currently a task registered for this key, get it.
+                    // Either way, store this new task as one for this key.
+                    Task<TResult>? prevTask;
+                    lock (s_tasks)
+                    {
+                        s_tasks.TryGetValue(key, out prevTask);
+                        s_tasks[key] = task;
+                    }
+
+                    // If there was a previous task, wait for it to complete, in order to ensure there's only ever
+                    // one work item in flight for this key at a time.
+                    if (prevTask is not null && !prevTask.IsCompleted)
+                    {
+                        await prevTask.ContinueWith(_ => { }, cancellationToken, TaskContinuationOptions.None, TaskScheduler.Default).ConfigureAwait(false);
```

Comment:
This is essentially an await with a CT right?

Reply:
It awaits in a way that won't throw an exception from the task but will be a cancelable await and will throw a cancellation exception if the token has cancellation requested.